[REBASE&FF] ReportStatusCode: Always cache protocol before usage

Javagedes opened the pull request:
## Description

This commit stops the library for querying the protocol database for the
status code protocol at the time of use of the library function. This
logic can result in a TPL inversion if the protocol has not yet been
located and the caller is running in a TPL higher then TPL_NOTIFY.

Instead, the protocol is now located during construction of the library
during driver startup. If the protocol is not found, then an on-protocol
install event is registered. A deconstructor is also added to close the
event in the case that the driver is unloaded before the protocol is
installed.

This PR is a rebase and fast forward as it includes a new test.

- [x] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested

1. The written unit test, which failed in the first commit, now passes after the second commit
2. Verified that the event callback is successfully executed in all drivers that run before the protocol is produced, including the DXE Core.
3. Verified the steps to reproduce specified in #1601 no longer reproduces.

## Integration Instructions

N/A

Javagedes committed:
MdeModulePkg: Add unit tests for DxeReportStatusCodeLib

This commit adds a UEFI_APPLICATION unit test for DxeReportStatusCodeLib
that ensures that the platform does not invert the TPL on an initial
call to ReportStatusCode functionality.

This functionality has the potential to invert the TPL due to the fact
that on an initial call to report status code functionality, if the
EFI_STATUS_CODE protocal has not been located, it will attempt to locate
it. LocateProtocol performs a tpl raise to EFI_NOTIFY. If the caller of
ReportStatusCode is in a TPL higher then EFI_NOTIFY, we will invert the
TPL and assert.

apop5 commented on MdeModulePkg/Library/DxeReportStatusCodeLib/ReportStatusCodeLib.c:
```diff
@@ -611,3 +611,69 @@ ReportDebugCodeEnabled (
+  IN VOID       *Context
+  )
+{
+  InternalGetReportStatusCode ();
```
Should the event get closed? I think gEfiStatusCodeRuntimeProtocolGuid is only installed once...

Javagedes replied:
Ah yes. Thanks. Updated :)

makubacki replied:
If there is a reinstall, are you sure you'd want to keep using the older protocol instance?

Javagedes replied:
As @apop5 mentioned, is this installed more than once? Or should I just be safe and do it anyway? Technically with the current implementation, it also only uses the first protocol found, regardless of future possible installations.

If that is the case, then I'd probably need to register the event callback even if we found a protocol during start up.

Javagedes replied:
Another benefit if we do it this way, is I can just remove the `InternalGetReportStatusCode` from all library functions, as we can just assume (is it safe to assume this?) that the protocol will be cached if it exists.

makubacki replied:
It is possible that it could be reinstalled simply because that's an option and we can't control all code that interacts with the protocol database out there.

The expectation is usually that what would you get from a current `LocateProtocol()` call is what would be used moving forward.

My understanding is that prior behavior would also freshly locate the protocol on a `ReportStatusCode()` invocation. If that's the case, it should continue to be supported.

makubacki replied:
Though, in the current implementation, it will still update the global through `InternalGetReportStatusCode()`, right?

Javagedes replied:
Hmm That is a good point. `InternalGetReportStatusCode` returns early if the cached protocol is already set. So this library does not currently ever update the protocol if a new one is registered.

I can update the event callback to update it regardless, if you wish. 

But I cannot update `InternalGetReportStatusCode` or we will try and re-locate the protocol every time, which will make us continue to hit the current TPL inversion error.

makubacki replied:
I'm a little concerned about the number of entries in the event database leaving this open too given it is linked against so many DXE modules. But I think that's technically the more robust behavior and I lean toward it if the reinstall case was previously supported to preserve that behavior.

makubacki replied:
Actually, looking at the current code, it does not locate when the global is non-null, so I think closing is fine since that's the precedent.

apop5 commented on MdeModulePkg/Library/DxeReportStatusCodeLib/ReportStatusCodeLib.c:
```diff
@@ -611,3 +611,69 @@ ReportDebugCodeEnabled (
+  InternalGetReportStatusCode ();
+
+  // If the Status Code Runtime Protocol is not found, register a notification event for when it is installed.
+  if ((mReportStatusCodeLibStatusCodeProtocol == NULL) && (gBS != NULL) && (gBS->CreateEvent != NULL) && (gBS->RegisterProtocolNotify != NULL)) {
```
if gBS is valid, then CreateEvent and RegisterProtocolNotify should also be valid

https://github.com/microsoft/mu_basecore/blob/release/202502/MdeModulePkg/Core/Dxe/DxeMain/DxeMain.c#L42

they are populated in the default structure (mBootServices )

Javagedes replied:
Fair enough! Updated :)

os-d reviewed (approved):
Can you take this to edk2, too?

makubacki commented on MdeModulePkg/Library/DxeReportStatusCodeLib/ReportStatusCodeLib.c:
```diff
@@ -611,3 +611,75 @@ ReportDebugCodeEnabled (
+**/
+EFI_STATUS
+EFIAPI
+DxeReportStatusCodeLibConstructor (
```
A convenience function sometimes used for this pattern is `EfiCreateProtocolNotifyEvent()` in UefiLib. I'm not necessarily suggesting that you use it, and you'd have to refactor some logic slightly to do so, but it's an option.

Javagedes replied:
This was a good idea and I tried it out after your suggestion, but one of the things it does is signal the event once. Since this library constructor is also called for the DXE_CORE, the event list is not fully set up and we end up asserting.

Due to this, I will be closing this without implementing.

makubacki commented on MdeModulePkg/Library/DxeReportStatusCodeLib/ReportStatusCodeLib.c:
```diff
@@ -611,3 +611,75 @@ ReportDebugCodeEnabled (
+      return EFI_SUCCESS;
+    }
+
+    Status = gBS->RegisterProtocolNotify (
```
Perhaps printing on error to make use of the return code would be useful here too.

Javagedes replied:
Added an error check and log for the register function also. Thanks!

makubacki commented on MdeModulePkg/Library/DxeReportStatusCodeLib/TestDxeReportStatusCodeLibTestApp.c:
```diff
@@ -0,0 +1,84 @@
+
+  Status = RunAllTestSuites (Framework);
+
+EXIT:
```
nit: Convention would usually have this be `Exit`

Javagedes replied:
Updated to `Exit`. Thanks!

makubacki commented on MdeModulePkg/Library/DxeReportStatusCodeLib/TestDxeReportStatusCodeLibTestApp.h:
```diff
@@ -0,0 +1,24 @@
+**/
+
+#ifndef _TEST_DXE_REPORT_STATUS_CODE_LIB_H_
+#define _TEST_DXE_REPORT_STATUS_CODE_LIB_H_
```
```suggestion
#ifndef TEST_DXE_REPORT_STATUS_CODE_LIB_H_
#define TEST_DXE_REPORT_STATUS_CODE_LIB_H_

```

Javagedes replied:
Updated. Thanks!

kuqin12 commented on MdeModulePkg/Library/DxeReportStatusCodeLib/ReportStatusCodeLib.c:
```diff
@@ -611,3 +611,80 @@ ReportDebugCodeEnabled (
+    Status = gBS->RegisterProtocolNotify (
+                    &gEfiStatusCodeRuntimeProtocolGuid,
+                    Event,
+                    &Registration
```
I think we might need to make this `Registration` global and unregister is in a destructor. This is in case the driver unloads itself, then we will need leave a dangling pointer in the protocol DB.

Javagedes replied:
Cached the `Event`, not the registration, so that I can call `gBS->CloseEvent(...)` in the deconstructor.

kuqin12 commented on MdeModulePkg/Library/DxeReportStatusCodeLib/ReportStatusCodeLib.c:
```diff
@@ -611,3 +611,80 @@ ReportDebugCodeEnabled (
+  if ((mReportStatusCodeLibStatusCodeProtocol == NULL) && (gBS != NULL)) {
+    Status = gBS->CreateEvent (
+                    EVT_NOTIFY_SIGNAL,
+                    TPL_CALLBACK,
```
nit: we probably can make this run at a higher level. But this is just to extend the coverage of RSC calls.

Javagedes replied:
It can run at any level up to TPL_NOTIFY. If you have a preference, I will update. Just let me know which one you want.

makubacki replied:
I assume the request was to prioritize setting the global here to preempt TPL_CALLBACK notifications.

kuqin12 commented on MdeModulePkg/Library/DxeReportStatusCodeLib/ReportStatusCodeLib.c:
We need to remove this locate call. Otherwise, a call before our callback will still run into the same issue?

Javagedes replied:
I mentioned this in the PR comment, but this change merely follows the requested change in the attached issue and does not fully stop the possibility of a TPL inversion.

As this PR mentions, it is still possible to have an inversion if a driver is running before any Status Code protocol has been globally installed, and calls `ReportStatusCode` while in a TPL higher than `TPL_NOTIFY`.

I'm more than happy to fundamentally change the logic of this to only set the global via the constructor & protocol install event, removing the LocateProtocol from `ReportStatusCode` all together.

If that is your preference, just let me know.

kuqin12 replied:
So the proposal in the issue is to remove the locate on the fly and move it to the constructor only, so that we do not run into the same issue. We are already changing the logic quite a bit and adding a brand new test, might as well fully prevent the inversion from happening.

Javagedes replied:
Sounds good. Seems like I misunderstood the proposal. I will make the changes and re request reviews :)

kuqin12 replied:
Thanks a lot!

Javagedes replied:
Per this conversation and offline conversation, I am updating this library to no longer attempt to get the report status code protocol (if not found) during a `ReportStatusCode` call. This will completely prevent the possibility of TPL inversion.

After changes, we do an initial `GetProtocol` in the constructor. If it fails to find the protocol, we instead create an on protocol event callback to cache the protocol.

The event will only ever be called once (uninstalling itself).

If the library is unloaded and the event has not been called, we will close the event.

kuqin12 commented on MdeModulePkg/Library/DxeReportStatusCodeLib/ReportStatusCodeLib.c:
```diff
@@ -611,3 +611,80 @@ ReportDebugCodeEnabled (
+  EFI_STATUS  Status;
+  VOID        *Registration;
+
+  InternalGetReportStatusCode ();
```
And instead, we need to move the locate here.

Javagedes replied:
`InternalGetReportStatusCode ();` is what does the locate. so this portion is already correct.

kuqin12 commented on MdeModulePkg/Library/DxeReportStatusCodeLib/TestDxeReportStatusCodeLibTestApp.c:
```diff
@@ -0,0 +1,84 @@
+/** @file
+  UEFI based application for unit testing the DxeReportStatusCodeLib.
```
I like the tests :)

kuqin12 commented on MdeModulePkg/MdeModulePkg.dsc:
```diff
@@ -351,6 +351,15 @@
+        UnitTestBootLib|UnitTestFrameworkPkg/Library/UnitTestBootLibNull/UnitTestBootLibNull.inf
+  }
+  # MU_CHANGE [END] - Add constructor to cache protocol or register for protocol notification
   MdeModulePkg/Library/RuntimeDxeReportStatusCodeLib/RuntimeDxeReportStatusCodeLib.inf
```
I think you will need to make the same change for this runtime instance.

Javagedes replied:
This library instance is the backing library for `DXE_CORE`, `DXE_DRIVER`, and `DXE_RUNTIME_DRIVER` (among multiple others), so this should suffice.

Javagedes commented:
I have re-requested all reviews, as I have made a fundamental change to the pull request.

Javagedes committed:
MdeModulePkg: DxeReportStatusCodeLib: Do not query protocol.

This commit stops the library for querying the protocol database for the
status code protocol at the time of use of the library function. This
logic can result in a TPL inversion if the protocol has not yet been
located and the caller is running in a TPL higher then TPL_NOTIFY.

Instead, the protocol is now located during construction of the library
during driver startup. If the protocol is not found, then an on-protocol
install event is registered. A deconstructor is also added to close the
event in the case that the driver is unloaded before the protocol is
installed.

Javagedes committed:
MdeModulePkg: RuntimeDxeReportStatusCodeLib: Do not query protocol.

This commit stops the library for querying the protocol database for the
status code protocol at the time of use of the library function. This
logic can result in a TPL inversion if the protocol has not yet been
located and the caller is running in a TPL higher then TPL_NOTIFY.

Instead, the protocol is now located during construction of the library
during driver startup. If the protocol is not found, then an on-protocol
install event is registered. A deconstructor is also added to close the
event in the case that the driver is unloaded before the protocol is
installed.
